### PR TITLE
Synthesize a `TokenStream` for `StmtKind::Empty`

### LINKED ISSUE #80760: Panic with 'Missing tokens for nt ";"' when passing statement to proc macro from declarative macro
This the the lib.rs of a proc macro crate `pm`:

```rust
use proc_macro::TokenStream;

#[proc_macro]
pub fn pm(input: TokenStream) -> TokenStream {
    // It works if this is `format!("{}", input)`
    format!("{:?}", input);
    TokenStream::new()
}
```

And here is the main.rs of another crate:

```rust
macro_rules! t {
    // It works if I take a tt
    ($s:stmt) => {
        pm::pm!($s);
    }
}

t!(;);
```

When building, this outputs:

```
error: proc macro panicked
 --> src/lib.rs:3:9
  |
3 |         pm::pm!($s);
  |         ^^^^^^^^^^^^
...
7 | t!(;);
  | ------ in this macro invocation
  |
  = help: message: Missing tokens at src/lib.rs:3:16: 3:18 (#6) for nt ";"
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error
```

### Version it worked on

1.49.0, 1.50.0-beta.5

### Version with regression

rustc 1.51.0-nightly (da305a2b0 2021-01-05)  
binary: rustc  
commit-hash: da305a2b00530aa34dea4e48389204c26fa35dbb  
commit-date: 2021-01-05  
host: x86_64-unknown-linux-gnu  
release: 1.51.0-nightly

### PR DESCRIPTION
**Pull request by Aaron1011**
Fixes #80760

### CODE REVIEW

**Comment by petrochenkov** [comment]
`StmtKind::Empty` is only for diagnostics they said, we only want to report a lint they said.
1.5 years pass and `;` is already a full-fledged statement accepted by `stmt` matcher, just through the power of natural bug evolution.

**Comment by petrochenkov** [comment]
>At least `#[attr] ;` is not supported yet.

Oh well.
```rust
macro_rules! m {
    ($stmt:stmt) => { #[allow(bad_style)] $stmt }
}

fn main() {
    m!(;);
}
```

**Comment by petrochenkov** [comment]
I can't complain about the end result, but the process here was funny.
Of course, to align with the token-based expansion model better `;` would *eventually* become a statement, with some lang team process and approval, but until then it was still supposed to be only a "filler" between other statements, a part of the block syntax.

File: compiler/rustc_parse/src/lib.rs [hunk 1]
```diff
@@ -257,7 +257,13 @@ pub fn nt_to_tokenstream(
     let tokens = match *nt {
         Nonterminal::NtItem(ref item) => prepend_attrs(sess, &item.attrs, nt, item.tokens.as_ref()),
         Nonterminal::NtBlock(ref block) => convert_tokens(block.tokens.as_ref()),
-        Nonterminal::NtStmt(ref stmt) => prepend_attrs(sess, stmt.attrs(), nt, stmt.tokens()),
+        Nonterminal::NtStmt(ref stmt) => {
+            if let ast::StmtKind::Empty = stmt.kind {
+                Some(tokenstream::TokenTree::token(token::Semi, stmt.span).into())
```
Comment (petrochenkov) [inline compiler/rustc_parse/src/lib.rs:262]: As the example in https://github.com/rust-lang/rust/pull/80789#issuecomment-756447607 shows, empty statements can have attributes too.
So `prepend_attrs` also needs to be used here in some form.

Comment (Aaron1011) [reply compiler/rustc_parse/src/lib.rs]: Fixed

**Comment by Aaron1011** [comment]
@petrochenkov: The current behavior w.r.t to attributes is very odd - any attributes added via `#[attr] $empty_stmt` are not actually invoked. I've added a test to ensure that we detect if this changes.

**Comment by petrochenkov** [comment]
r=me after moving the common part `prepend_attrs(sess, stmt.attrs(), nt, ...)` to outside of the condition.

**Pull request merged**